src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -154,202 +110,179 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
-            // the connection was starting
-            if (ChangeState(from: ConnectionState.Connecting, to: ConnectionState.Connected) == ConnectionState.Connecting)
+        private async Task SelectAndStartTransport(TransferFormat transferFormat)
+        {
+            if (_requestedTransportType == TransportType.WebSockets)
             {
-                _closeTcs = new TaskCompletionSource<object>();
-
-                Input.OnWriterCompleted(async (exception, state) =>
-                {
-                    // Grab the exception and then clear it.
-                    // See comment at AbortAsync for more discussion on the thread-safety
-                    // StartAsync can't be called until the ChangeState below, so we're OK.
-                    var abortException = _abortException;
-                    _abortException = null;
-
-                    // There is an inherent race between receive and close. Removing the last message from the channel
-                    // makes Input.Completion task completed and runs this continuation. We need to await _receiveLoopTask
-                    // to make sure that the message removed from the channel is processed before we drain the queue.
-                    // There is a short window between we start the channel and assign the _receiveLoopTask a value.
-                    // To make sure that _receiveLoopTask can be awaited (i.e. is not null) we need to await _startTask.
-                    Log.ProcessRemainingMessages(_logger);
+                Log.StartingTransport(_logger, _requestedTransportType, Url);
+                await StartTransport(Url, _requestedTransportType, transferFormat);
+            }
+            else
+            {
+                var negotiationResponse = await GetNegotiationResponse();
 
-                    await _startTcs.Task;
-                    await _receiveLoopTask;
+                // If we get disposed while starting, stop ASAP.
+                CheckDisposed();
 
-                    Log.DrainEvents(_logger);
+                // This should only need to happen once
+                var connectUrl = CreateConnectUrl(Url, negotiationResponse.ConnectionId);
 
-                    await Task.WhenAny(_eventQueue.Drain().NoThrow(), Task.Delay(_eventQueueDrainTimeout));
+                // We're going to search for the transfer format as a string because we don't want to parse
+                // all the transfer formats in the negotiation response, and we want to allow transfer formats
+                // we don't understand in the negotiate response.
+                var transferFormatString = transferFormat.ToString();
 
-                    Log.CompleteClosed(_logger);
-                    _logScope.ConnectionId = null;
+                foreach (var transport in negotiationResponse.AvailableTransports)
+                {
+                    // If we get disposed while starting, stop ASAP.
```

Comment:
Why?

Reply:
I had a version of this that worked without the lock originally. I removed that and didn't tidy these up.